backend/tools/event_sender/src/main.rs

```diff
@@ -115,40 +143,61 @@ async fn app() -> anyhow::Result<()> {
     };
 
     let target = cli.count * cli.burst;
-    let sleep_duration = Duration::from_millis(cli.latency as u64);
+
+    // Random number generator used to product random intervals
+    fn random_num(from: usize, to: usize) -> usize {
+        let mut rng = rand::thread_rng();
+        rng.gen_range(from..to)
+    }
 
     println!(
-        "Sending out ~{} events, {} events every {}ms interval",
-        target, cli.burst, cli.latency
+        "\nSending out a maximum of {} events, over a maximum duration of {} seconds. \nProcess will stop at whichever comes first. \nEvents sent at random intervals between {}-{}ms. \n",
+        target, (cli.total_duration / 1000), cli.min_latency, cli.latency_max
     );
 
     let mut sent = 0;
+    // Used for a time reference staring point
+    let start = Instant::now();
+    // Decalred as mut in order to allow the value to change
+    let mut duration = start.elapsed();
+
     for _ in 0..(target) {
-        let mut taken = 0;
-        for ev in (&mut iter).take(cli.burst) {
-            let _ = channel_a
-                .basic_publish(
-                    cli.exchange.as_str(),
-                    cli.routing_key.as_str(),
-                    BasicPublishOptions::default(),
-                    ev.as_slice(),
-                    BasicProperties::default(),
-                )
-                .await?
-                .await?;
-            taken += 1;
-        }
-        if cli.burst > taken {
-            println!("Exhausted source, stopping early!");
-            break;
-        }
-        sent += taken;
-        if sleep_duration.as_millis() > 0 {
-            tokio::time::sleep(sleep_duration).await;
+        // Loop until the elapsed time has reached the calculated total duration or event count has been reached
+        while duration.as_millis() < cli.total_duration.try_into().unwrap() && sent < cli.count {
```

Comment:
this while loop is inside the `for _ in 0..(target)` loop, even if it works it is not correct logic

Reply:
Of course, not sure why I have placed it within the for loop! Will update. Thanks!